core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala

```diff
@@ -694,7 +694,7 @@ private[storage] class PartiallyUnrolledIterator[T](
   }
 
   override def next(): T = {
-    if (unrolled == null) {
+    if (unrolled == null || !unrolled.hasNext) {
```

Comment:
Isnt this a correctness issue though? If unrolled has no more elements the correct result is to return from the other iterator not throw an exception.